### ISSUE TITLE
Freeze with Analytics + Messaging in IL2CPP/.NET4.X in Unity 2018.2

### ISSUE THREAD

**Issue by BenSteenson** [firebase/quickstart-unity#184]
Unity Version: 2018.2.0f2
Build Device: Windows 10 64-bit
Android Test Device: Sony Xperia ZX (Android 8.0)
JDK: 1.8.0u172
NDK: r13b
Firebase: 1.1.1

Build Configuration:
- Min API Level: 21
- Target API Level: Automatic (27)
- Scripting Runtime Version: .NET 4.X Equivilant
- Scripting Backend: IL2CPP
- API Compatibility Level: .NET 4.X
- C++ Compiler Configuration: Release
- Target Architecture: ARMv7
- Strip Engine Code: On
- Build System: Gradle
- Minify: None

Whenever building a specific project with Firebase in **Release Mode**, the app will freeze with no error messages, logs or warnings in logcat, and no ANRs once it finishes the splash screen. When the app is set to Debug Mode, or Firebase is removed from the project, the application runs fine (albeit slower in Debug mode).

The Unity project uses RestSharp and makes heavy use of async/await for networking purposes. When these are removed, the app seems to work in Release mode (although cripples functionality as it's core the app).

I have made sure to remove the 2.0 Parse DLLs and enable the 4.X Parse DLLs for all platforms. I have Updated and Force Resolved the dependencies and enabled Auto-Resolution, Explode AARs and Install Android Packages in the Google Play Resolver settings.

The following is the Initialization logic for Firebase on a GameObject that exists in the scene. 

```csharp
private async void Start()
{
	await FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task =>
	{
		DependencyStatus dependencyStatus = task.Result;
		if (dependencyStatus == DependencyStatus.Available)
		{
			FirebaseAnalytics.SetAnalyticsCollectionEnabled(true);
			FirebaseAnalytics.SetMinimumSessionDuration(new System.TimeSpan(0, 0, 10));
			FirebaseAnalytics.SetSessionTimeoutDuration(new System.TimeSpan(0, 30, 0));
		}
		else
		{
			Debug.LogError("Could not resolve all Firebase dependencies: " + dependencyStatus);
		}
	});

	//Create and Run the gameobjects for our app
	Launch();
}
```

It will also break if I remove the Await and simply use a boolean flag to check if the app is initialized on Update/main thread.

**Comment by Woody4618** [comment]
I am having the same problem. 
Unity 2018.2 f2
Messaging and Analytics enabled. 
Mono scripting backend 
.net 3.5
Min api level 19 
Game freezes after splash screen without any errors. 
Once it was loading and then crashed when recieving a push notification afterwards it kept freezing on start again.

**Comment by chrisfairc** [comment]
I am seeing my app freeze on android with the latest firebase SDK in Unity. I updated my android platform and android studio to the latest too. 

The project is on Unity 2017.1.5. Only firebase analytics is in the project. The app freezes on startup. This is only with builds made recently.

**Comment by BenSteenson** [comment]
I'm very, very certain this is a fault with the ParseSDK. Commenting out the _await FirebaseApp.CheckAndFixDependenciesAsync()_ line and any other Firebase-related code allowed the app to run correctly, even with the plugin installed. When I ran some Debug.Logs, the application faulted just after it completed the Firebase initialization.

**Comment by stewartmiles** [comment]
@alexames could you take a look?

**Comment by alexames** [comment]
Will do.

**Comment by alexames** [comment]
I think the issue here is the async/await. If you remove those, does this start working?

**Comment by BenSteenson** [comment]
I made the following code changes (the base script itself has also evolved over the time, hence the formatting changes):

```
        /// <summary>
        /// Gets if Firebase is enabled and available for use in the application.
        /// </summary>
        public static bool FirebaseActive { get; private set; } = false;

        private void Start()
        {
            //Firebase needs to initialise first, it will then initiate the rest of the launch process

            if (UseFirebase)
            {
                FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task =>
                {
                    if (task.Result == DependencyStatus.Available)
                    {
                        FirebaseAnalytics.SetAnalyticsCollectionEnabled(true);
                        FirebaseAnalytics.SetMinimumSessionDuration(new System.TimeSpan(0, 0, 10));
                        FirebaseAnalytics.SetSessionTimeoutDuration(new System.TimeSpan(0, 30, 0));
                        FirebaseActive = true;
                    }
                    else Debug.LogError($"Could not resolve all Firebase dependencies: {task.Result.ToString()}");

                    ConfigureApp();
                }, System.Threading.Tasks.TaskScheduler.FromCurrentSynchronizationContext());
            }
            else ConfigureApp();
        }
```

I was able to build 3 builds in Release mode with the app working correctly. I then updated Firebase to 5.2.1, UAP to 1.4 and the Incremental Compiler to 0.0.42-preview19, and it stopped worked.

I will verify if this is an issue with the Incremental Compiler shortly.

**Comment by BenSteenson** [comment]
On my development machine, I updated Unity to 2018.2.4f1, manually deleted some left over files from Firebase 5.0.0 and 5.2.0, removed permissions and the package name from my main AndroidManifest.xml and turned on Minify. When I pulled the changes on my build machine (after having deleted some cache folders in the Library folder), the build came out perfectly.

This is def. due to using async on CheckAndFixDependenciesAsyncy() with ContinueWith(), and I would think the most recent fault was due to the update process for Firebase 5.2.1 (which is probably on me). I'll re-open if it occurs again.
